### DISCUSSION TITLE
[FEATURE] Support for MSVC

### DISCUSSION THREAD

**Discussion by jfalcou** [jfalcou/eve#1550] [Done & Done!]
This is a mutable, ongoing issue to discuss/list steps for proper MSVC support.

# Current Status
Since MSVC 2022 19.30, C++20 support has drastically improved. Building blocks libraries like kumi and raberu now natively compile on MSVC without any hassle. This has triggered the investigation of how much EVE needs to be fixed to make MSVC EVE a thing.

Note that we currently don't care about code generation quality as our first goal here is to get the library to compile.

# Plan of Actions
We need to use our module based tests system to validate EVE code on MSVC independant piece after the next.
This is the ordered list of module to make pass:

- [x] `unit.api.exe`
- [x] `unit.memory.exe`
- [x] `unit.core.exe`
- [x] `unit.math.exe`
- [x] `unit.algo.exe`
- [x] `unit.views.exe`
- [x] All other modules
- [x] `doc.exe`

Best process so far: 
- Hack away on MSVC then push to validate CI.
- If  anything break on g++/clang, fix accordingly then rince & repeat.
- Prefer code reorganisation/rewrite to  compiler specific `#if #else #endif` whenever possible

MSVC also produces a lot of obnoxious warning for conversions. Some are really stupid IMHO but try to fix them
as best as we can unless it makes the code ugly.

# Current Status
- [X] `unit.arch.exe` compiles, pass tests and validate CI
- [X] `unit.meta.exe` compiles and pass tests and validate CI
- [X] `unit.internals.exe` compiles and pass tests and validate CI

**Comment by DenisYaroshevskiy** [comment]
CI?

I'll have to install windows somewhere, won't I?

**Comment by jfalcou** [comment]
Was thinking of using appveyor

**Comment by Tradias** [comment]
Is there anything you need help with here?

Apart from appveyor, I have only ever used Github Actions and they were easy to set up.

**Comment by jfalcou** [comment]
I just need to find time to tackle the stupid errors that prevent other test to compile.

 Appveyor is setup already.

**Comment by jfalcou** [comment]
It looks like some compile time traits we have don't fly properly but the error messages are outrageous.

I will push a more up-to-date list of breaking thing later today.

**Comment by jfalcou** [comment]
So mostly we need help understanding why msvc don't understand our properly standard templatr monstrosity

**Comment by Tradias** [comment]
Sure, sounds like something I can investigate.

**Comment by jfalcou** [comment]
Good. Let me update this issue later withore info

**Comment by jfalcou** [comment]
@Tradias added a small roadmap on how to tackle the thing. Tell me if you need help.

**Comment by jfalcou** [comment]
# unit.api targets
 - Looks like we try to use constepxr value as lambda capture in some of the shuffle tests and MSVC doesn't appreciate that

**Comment by Tradias** [comment]
Looks good, I have managed to get `unit.core.exe` to compile without significant code changes. Creating a pr now.

**Comment by jfalcou** [comment]
Good job wow :D
Did you used the in flight PR that fixes unit.internals (https://github.com/jfalcou/eve/pull/1458) ? or were the issue unrelated ?

**Comment by Tradias** [comment]
I didn't know about https://github.com/jfalcou/eve/pull/1458, one of the issue was in fact related. I think my solution might be faster to compile but potentially incorrect.

**Comment by jfalcou** [comment]
I had the same code at one point but it failed on other compilers. Maybe we can borrow your cocnept based strategy but lay  it out like ours ?

**Comment by Tradias** [comment]
Yes that is possible. I have reverted it to your version for now.

**Comment by jfalcou** [comment]
Merged the CI adaptation to see if it is faster to compiles

**Comment by ThomasRetornaz** [comment]
May i could help? (I'm currently integrate kiwaku +eve for an image processing project and i will really appreciate msvc support for EVE) 
I see two unmerged PR regarding msvc support and two dedicated branch
Does it make sens to rebase the branch or its better to restart a fresh tria from maste
Regards
TRl

**Comment by Nxirda** [comment]
# State 
Every "main" part is compiling.  
The main blocking factor  was the way MSVC is handling `EBO`. 
In most places, special cases for MSVC handled via `#ifdefs` have been removed.

## Notes : 
- Lambda captures of constexpr variable is not implicit
- Prefer explicit template parameter of lambda with parameter types containing NTTPs (ex : `std::integral_constant`) 
- We currently make no guarantees about the qualify of the codegen

## Flags : 
Quite a few warnings are disabled, a bunch of which are actually meaningless as of now. 

Relevant MSVC warnings are : 
- /wd4244 : conversion to smaller integer type
- /wd4267 : conversion from size_t to smaller integer type
- /wd4146 : unary minus on unsigned (IIRC used as some trick somewhere/in printing)
- /wd4018 : signed/unsigned comparison without explicit cast
- /wd4305 : truncation conversion
- /wd4334 :  result of 32-bit shift implicitly converted to 64 bit

These should be fixed in a proper PR addressing these for every compiler.

## Remaining parts :
- [x] `unit.memory.exe`
- [x] `unit.algo.exe`
- [x] `unit.views.exe`
- [x] `doc.exe`

The main issue seems to arise from the fact that MSVC iterators are not naturally convertible to pointers meaning we should use `.data()` and `.size()` whenever a pointer is needed.

## Current Status that compiles, pass tests and validate CI
- [X] `unit.arch.exe` 
- [X] `unit.meta.exe`
- [X] `unit.internals.exe`
- [X] `unit.api.exe`
- [X] `unit.core.exe`
- [X] `unit.math.exe`

## Clang-CL 
We tried to support Clang-CL in the meantime there are two issues that arose immediately. All tests where we somehow encounter the old callable systems seems to break. We should fix this before investigating further. PCH support seems incomplete on their side requiring `-DEVE_USE_PCH=OFF` to have a build that ~works.
